### DISCUSSION TITLE
PyBullet: Programmatically set joints position and limit with createMultiBody

### DISCUSSION THREAD

**Discussion by ycps** [bulletphysics/bullet3#3779] [General]
Hi, I'm experimenting with randomized robot generation, but could not find a way to set joints position or limit with `createMultiBody`: joints seem to always be created at the center of shapes without range limits.

I understand that using URDF or the other importers is the expected way to create robots, but I don't know what is the best approach when most of the robot features should be randomized.

I searched about this, but could only find this similar questions in the PyBullet forum:
- [Programmatically set the revolute (hinge) joint limits?](https://pybullet.org/Bullet/phpBB3/viewtopic.php?f=24&t=12927&sid=00c3f4ea457563f36e1e3883eeaa48cb)
- [how to change joint position](https://pybullet.org/Bullet/phpBB3/viewtopic.php?t=11953)

Thank you for Bullet!

**Comment by erwincoumans** [comment]
Joint position can be set using the linkPositions argument (and then change joint angle using resetJointState)
See this example:
https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/examples/createMultiBodyLinks.py

Joint limits are indeed not exposed yet. I think the joint limit constraints are actually added but not activated. We should implement a 'changeDynamics' method to modify the joint limits.

**Comment by erwincoumans** [comment]
Here is a CL that allows to enable/modify lower/upper joint limits and also joint limit max force.
See https://github.com/bulletphysics/bullet3/pull/2919
Note that the CL doesn't update the internal info, so the joint lower/upper limits in getJointInfo will not be updated!
